Add security considerations about foreign claims (#185)
spiffe/spiffe: standards/SPIFFE_Federation.md

Finally, it should be noted that the "links" in this chain are formed by individual HTTPS request operations (against the various bundle endpoint servers in the chain), and these operations are likely occurring at different times. SPIFFE bundle endpoint clients should be logging these HTTPS request operations, and administrators should take care to preserve these logs for future forensic analysis if necessary.

### 7.7 Processing claims asserted by foreign trust domains

> SPIFFE Federation allows workloads in one trust domain to authenticate SVIDs, and all the claims therein, issued by other trust domains. By an "authentic" claim, we mean specifically that it can be safely assumed that the foreign trust domain authority _asserted_ the claim. This is not the same thing as saying that the claim is _true_, or that it is _correctly interpreted_.

Reviewer: ```suggestion
SPIFFE Federation allows workloads in one trust domain to authenticate SVIDs, and all the claims therein, issued by other trust domains. An "authenticated" claim means specifically that it can be safely assumed that the foreign trust domain authority _asserted_ the claim. This is not the same thing as saying that the claim was _true_ when asserted by the authority, or that it will be _correctly interpreted_.
```

Middle sentence still needs some work, I ran out of steam.

Author: I'm curious about authentic -> authenticated, @justinburke , can you say something about why you're proposing that change?

Reviewer: First, to me, `authentic claim` primarily suggests that the claim is credible, is true. Whereas, to me, `authenticated claim` more clearer suggests that the source of the claim (the foreign trust domain authority, in this case) did in fact assert the claim, irrespective of the veracity of the claim itself.  (I'm probably splitting hairs... but I put suggestion in because this phrasing interrupted my reading flow.)

Second, this is the only occurrence (at least, in this PR) of the phrase `authentic claim`.  If we're going to ask the reader to accept some definition of a phrase, I feel that we should make use of that phrase.

Author: In my mind, the verb authenticate means to prove something is authentic. Elsewhere in this document we use authenticated as a past-tense verb, not as an adjective.

"authenticated claim" is also not used anywhere else in the document, and the intention is to sharply define "authentic" rather than "authentic claim".

> SPIFFE Federation allows workloads in one trust domain to authenticate SVIDs, and all the claims therein, issued by other trust domains.  By an "authentic" claim, we mean specifically that it can be safely assumed that the foreign trust domain authority _asserted_ the claim.  This is not the same thing as saying that the claim is _true_, or that it is _correctly interpreted_.

Reviewer 2: 💯

> Operators should use caution and good judgement in evaluating if and how they should process claims asserted in SVIDs from foreign trust domains.

Reviewer 2: Should we nudge folks in the direction of allowlisting, operator-driven agreement, etc? Here's a poorly-thought-through example of what that could look like

```suggestion
Operators should use caution and good judgement in evaluating if and how they should process claims asserted in SVIDs from foreign trust domains. At a minimum, the claims to be utilized and their exact meaning should be agreed upon amongst the operators of all involved trust domains beforehand. Unless otherwise agreed upon, it is strongly recommended that claims from foreign trust domains not be processed.
```

Author: Added my own attempt at this, PTAL

> Operators should use caution and good judgement in evaluating if and how they should process claims asserted in SVIDs from foreign trust domains. The recommended posture is not to process foreign claims that have not been evaluated accoring to the framework described here (i.e. default deny), then assert exceptions for specific claims from specific trust domains to be allowed if they pass muster.

Reviewer: ```suggestion
Operators should use caution and good judgement in evaluating if and how they should process claims asserted in SVIDs from foreign trust domains. The recommended posture is to process foreign claims only after they have been evaluated according to the framework described below; for example, start with a default deny policy, then define exceptions for specific claims from specific trust domains to be allowed if they pass muster.
```

Reviewer: What does "pass muster" mean in this context?

Author: It means that, after being evaluated as described, they survive that scrutiny and the operator approves of processing them.

Maybe we just cut that phrase out

"The recommended posture is to process foreign claims only after they have been evaluated according to the framework described below; for example, start with a default deny policy, then define exceptions for specific claims from specific trust domains."

I guess it's clear from the first clause of the sentence what the criteria is for an exception to the default deny?

> It bears repeating that implementations should never process any claims from SVIDs that fail authentication (except in a forensic context).

Reviewer: ```suggestion
Implementations are strongly advised to process claims contained in an SVID only after successfully authenticating the SVID; claims contained in unauthenticated SVIDs should be not be processed.
```

#### 7.7.1 SVID Claims

> For the purpose of this discussion, a *claim* is an assertion about some attribute of the subject of an SVID, or about the SVID itself. The canonical SPIFFE claim is the SPIFFE ID, which is a claim about the subject of the SVID. An example of a claim about the SVID itself is the expiration time.

Reviewer: ```suggestion
For the purpose of this discussion, a *claim* is an assertion about some attribute of the subject of an SVID, or about the SVID itself. The canonical SPIFFE claim is the "SPIFFE ID", which is a claim about the subject of the SVID. An example of a claim about the SVID itself is the expiration time.
```

> SVID formats (to date including X.509 and JWT) afford flexibility to encode additional claims, including claims that have not been evaluated by any standards body.

Reviewer: What is an example of a claim that has been evaluated by a standards body?

I think what this paragraph is trying to say is:
SVID formats (e.g. X.509 and JWT) intentionally include the ability to include claims which are not specified by the SVID or underlying format (e.g. X.509, JWT) specifications.

Author: No, here I'm thinking of things like X.520 that define standard attributes https://www.itu.int/rec/T-REC-X.520-201610-S

These can be encoded in X.509 SANs as General Names.

Reviewer: Suggested edit from last week's SIG-Spec discussion: For example, X.509 specification, the DNS SAN attribute may be set even though the X509-SVID specification does not define its use. Contrarily, X.509-SVID specification allows for the use of custom SAN extensions which have not been defined by either the X509-SVID or X.509 specifications.

In considering how to process claims made by foreign trust domains, it is helpful to consider three categories of claims:

  1. Claims required by and/or defined by the SPIFFE specifications for the SVID type
  2. Claims that are standardized by the relevant standards body for the document format (i.e. ITU-T for X.509, IETF for JWT)
  3. Non-standardized claims

Claims in the first category are well-defined in their meaning, and in many cases implementations are required to process them to adhere to SPIFFE specifications.

Claims in the second category have standardized meaning, but operators have some choice whether to and how to process them.

> Claims in the third category are not standarized, so operators need to consider them on a case-by-case basis, possibly including navigating different meanings by different trust domains. Needless to say, this category should be approached with the highest degree of caution.

Reviewer: ```suggestion
Claims in the third category are (by definition) not standardized, so operators need to consider them on a case-by-case basis, possibly including navigating different meanings by different trust domains. This category should be approached with the highest degree of caution.
```

Reviewer: "different meanings by different trust domains" is a terrific point to make. I hope that there is an example or more text about this below.

#### 7.7.2 Claim scope

> Some claims, like the SPIFFE ID are _explicitly_ scoped to a trust domain by their structure. Because SVIDs are always issued by a particular trust domain authority, _all_ the claims therein are _implicitly_ scoped to that trust domain.  This scoping should be preserved when presenting claims to any decision-making process.

Reviewer 2: 💯

> Some claims, like the SPIFFE ID are _explicitly_ scoped to a trust domain by their structure. Because SVIDs are always issued by a particular trust domain authority, _all_ the claims therein are _implicitly_ scoped to that trust domain. This scoping should be preserved when presenting claims to any decision-making process.

Reviewer: ```suggestion
Some claims, like the SPIFFE ID, are _explicitly_ scoped to a trust domain (i.e. the claim explicitly includes a trust domain). _All_ claims contained within an SVID, which do not contain an explicit scope, are assumed to be _implicitly_ scoped to the trust domain of the SPIFFE ID; this scoping should be preserved when presenting claims to any decision-making process.
```

Note that the issuer of an SVID need not be a member of the trust domain of the SVID, or any trust domain at all. In the original text, "that trust domain" appear to refer to the trust domain of the "trust domain authority", which may or may not be the same as the SPIFFE ID in the SVID, or may not be any trust domain.

Author: I guess here I was thinking that the signing authority is always an "authorized authority" for the trust domain of the SVID by virtue of the (trust domain, bundle) tuple, even if it is not itself a member of the trust domain.  But, that's probably a little less clear than the language you have.

Author: I do have some concerns about the idea that claims that contain an explicit scope are exempt from scoping to the trust domain.

For example, consider a claim that encodes a Kubernetes pod `<namespace>/<name>`.  I think the `<namespace>` part of this claim is arguably _an explicit scope_, but I still think the claim is implicitly scoped to the SPIFFE trust domain, and should not be exempt from having that scoping preserved when presenting the claim to a decision-making process. If anything, the scope is the intersection of the trust domain and any other explicit scope.

Reviewer: During last week's SIG-Spec meeting, we discussed a bit. IIRC, we generally thought that it would be helpful to highlight potential confused deputy problems when processing foreign claims.

Based on Spike's comments above and SIG-Spec discussion, I no longer think that we should suggest an implicit scoping. However, to help clarify the potential security problem (confused deputy), we could consider integrating this phrase from the discussion: "Qualified by the trust domain name".
So the emphasis isn't on "scoping" or "namespaces", but on "qualification", as in "foreign trust domain X has asserted claim Y", where claim Y may contain a reference to yet another trust domain.

As a hypothetical example, consider a claim about the version number of the control plane software that generated the SVID.  Different trust domain control planes might use entirely different software packages, and as such the version numbers are not directly comparable in an unscoped context.

#### 7.7.3 Interpretation of claims

> If claims are to be processed in making a security decision (such as whether to allow a request or connection), it is critical that the authority issuing the claim has the same meaning or interpretation of the claim as the entity making the decision.

Reviewer 2: I think it is worth calling out here that there is some level of trust we place in the authority to make the assertion.

```suggestion
If claims are to be processed in making a security decision (such as whether to allow a request or connection), it is critical that the authority issuing the claim has the same meaning or interpretation of the claim as the entity making the decision, and that the validator trusts the authority to make such an assertion faithfully.
```

Author: Maybe I'm misunderstanding what you mean by "faithfully"

If you're referring to whether the authority makes claims that are true, then I think this is covered in the next section.

In considering the categories of claims noted in [Section 7.7.1](#771-svid-claims), these should be treated with an increasing level of scrutiny around interpretation.

Also consider exactly who is operating the trust domain making the claim.  Trust domain operators may be able to agree on the meaning of a particular claim.  For example, if an organization operates multiple trust domains for administrative reasons, it may be straightforward to agree on an interpretation of claims within that organization (even if the claims are not publically standardized).  The operator should, however, limit processing of that claim only to the trust domains with which the meaning has been agreed.

#### 7.7.4 Veracity of claims

> Just because a trust domain authority asserts a claim does not gaurantee the claim is true.  When choosing whether to accept particular claims from particular trust domains, operators may wish to conduct audits, ask for evidence, or otherwise investigate the processes by which the trust domain makes the claim.

Reviewer 2: 💯

> Just because a trust domain authority asserts a claim does not gaurantee the claim is true.  When choosing whether to accept particular claims from particular trust domains, operators may wish to conduct audits, ask for evidence, or otherwise investigate the processes by which the trust domain makes the claim.

Reviewer 3: `gaurantee` ==> `guarantee`

> Just because a trust domain authority asserts a claim does not guarantee the claim is true. When choosing whether to accept particular claims from particular trust domains, operators may wish to conduct audits, ask for evidence, or otherwise investigate the processes by which the trust domain makes the claim.

Reviewer: ```suggestion
Just because a trust domain authority asserts a claim does not guarantee the claim is true at time of evaluation or was true at the time that the authority asserted the claim. When choosing whether to accept particular claims from particular trust domains, operators may wish to conduct audits, ask for evidence, or otherwise investigate the processes by which the trust domain makes the claim.
```

Author: @justinburke can you talk about why it's important to highlight the difference between the question "is the claim true" and "was the claim true when it was asserted"?  Like, is there some specific concern about the temporal question with relation to foreign trust domains?

Reviewer 2: This is exactly one of the sharp edges that drives me (personally) to discourage folks from embedding subject claims in an SVID. Often times, the claims that are asserted are not natural properties, and as such they can change. This has famously been a challenge for human-oriented X.509 auth wherein security clearance, team membership, etc is embedded and used for authorization. For these kinds of details, it is always better (IMO) to delegate the lookup to a secondary system.

I can see a few ways that this issue occurs for workloads too. One contrived example could be an assertion that the workload is running in/on a specific location or host, but the key for the SVID was not bound to the host and the workload has since been live migrated elsewhere.

Author: @evan2645 what you're talking about seems to be a sharp edge related to embedding extra claims at all.  That is to say, it's not a special concern related to the fact that you're interacting with a foreign trust domain.

I think a security considerations discussion of those kinds of issues fits more naturally in one or both of the SVID specs.

Reviewer: The suggested rephrasing that we discussing in SIG-Spec last week: "is true at the time of evaluation or was true at the time of issuance"

Author: This section is discussing whether the foreign trust domain can be trusted to tell the truth.

Introducing the concept of the time delay between issuance and evaluation here, without actually developing it, just serves to muddy the waters, rather than clarify.  I agree this is an important security consideration, but it really needs its own section where it can be given the attention it deserves.

> Imagine a hypothetical claim involving whether the peer workload is in a PCI-compliant (Payment Card Industry security specification, a standard for processing credit card data) environment. If we are going to share credit card data with workloads that are PCI-compliant, we might ask to see evidence of a 3rd party PCI audit and ask how the trust domain authority determines which workloads to issue the claim to. 

Reviewer: ```suggestion
Imagine a hypothetical claim involving whether the peer workload is in a PCI-compliant  environment (Payment Card Industry security specification, a standard for processing credit card data). Prior to sharing sharing credit card data with workloads that are PCI-compliant, we might ask to see evidence of a 3rd party PCI audit and ask how the trust domain authority determines which workloads to issue the claim to. In particular, the existence of a claim indicating PCI-compliance is merely a signal which has been asserted by a trust domain authority.
```

> If we determine we cannot trust core SPIFFE claims, like the SPIFFE ID, from a particular trust domain, it makes little sense to allow federation at all.  Non-required claims can be filtered by control planes and/or workloads if they are not trusted.

Reviewer: ```suggestion
If an SVID is issued by an untrusted or unknown authority, federation with the foreign trust domain has not been established and any claims contained within the SVID should be treated as untrusted. Once federation has been established with a foreign trust domain and SVIDs of that domain can be authenticated, control plane operators must determine whether to trust any claims which are not required by the SPIFFE specifications or underlying document format (e.g. X-509 or JWT specifications); non-standardized claims may be filtered by control planes and/or workloads if the issuer is not trusted to assert the specific claims.
```

Author: I think this rewording glosses over the point that it's actually worth asking yourself whether a foreign trust domain can be trusted to faithfully assert the SPIFFE-required claims, and that if the answer is no, you should not federate at all.
